code/globalincs/linklist.h

```diff
@@ -83,4 +83,87 @@ do {												\
 #define NOT_EMPTY(head)		((head)->next != (head))
 #define EMPTY(head)			((head) == nullptr || (head)->next == (head))
 
+template <class T>
+class volition_linked_list
+{
+	T* head;
+
+public:
+	class iterator
+	{
+		T* ptr;
+
+	public:
+		iterator(T* x)
+			: ptr(x)
+		{}
+
+		iterator(const iterator& it)
+			: ptr(it.ptr)
+		{}
+
+		iterator& operator++()
+		{
+			ptr = GET_NEXT(ptr);
+			return *this;
+		}
+
+		iterator operator++(int)
+		{
+			iterator tmp(*this);
+			operator++();
+			return tmp;
+		}
+
+		iterator& operator--()
+		{
+			ptr = GET_PREV(ptr);
+			return *this;
+		}
+
+		iterator operator--(int)
+		{
+			iterator tmp(*this);
+			operator--();
+			return tmp;
+		}
+
+		bool operator==(const iterator& rhs) const
+		{
+			return ptr == rhs.ptr;
+
+		}
+
+		bool operator!=(const iterator& rhs) const
+		{
+			return ptr != rhs.ptr;
+		}
+
+		T&& operator*()
```

Comment:
I would suggest returning a `const T&` here. Moving the pointer value may change the list we are iterating on.

Reply:
updated with a `const_iterator`